src/mcx_core.cu

```diff
@@ -1385,6 +1385,60 @@ __device__ inline int launchnewphoton(MCXpos* p, MCXdir* v, Stokes* s, MCXtime*
                     canfocus = (gcfg->srctype == MCX_SRC_SLIT);
                     break;
                 }
+                case (MCX_SRC_RING): {
+                    /* ring light source, can be a complete ring or a truncated ring. the expected parameters for definition of the
+                    light source are as follows:
+                    gcfg->srcparam1.x: minimum ring radius in mm
+                    gcfg->srcparam1.y: maximum ring radius in mm
+                    gcfg->srcparam1.z: first angular range in degrees. Defines the angular width of the first ring section
+                    gcfg->srcparam1.w: second angular range in degrees. Defines the angular width of the second ring section
+                    gcfg->srcparam2.x: full beam divergence in degrees. 0 means photons are launched straight down. angle greater than
+                    0 means that the photons are launched with a direction sampled from a normal distribution with FWHM of half the
+                    beam divergence.
+                    for simulating a complete ring the parameters should be defined as: gcfg->srcparam1.z = 180 and gcfg->srcparam1.w = 180
+                    */
+                    float minimum_radius = gcfg->srcparam1.x;
+                    float maximum_radius = gcfg->srcparam1.y;
+                    float cut_angle_1 = (TWO_PI / 360.f) * gcfg->srcparam1.z;
+                    float cut_angle_2 = (TWO_PI / 360.f) * gcfg->srcparam1.w;
+                    float random_radius = rand_uniform01(t);
+                    float random_angle = rand_uniform01(t);
+                    float angle_range_1[2] = {((TWO_PI / 2.f) - cut_angle_1) / 2.f, ((TWO_PI / 2.f) + cut_angle_1) / 2.f};
+                    float angle_range_2[2] = {(TWO_PI + (TWO_PI / 2.f) - cut_angle_2) / 2.f,
+                                              (TWO_PI + (TWO_PI / 2.f) + cut_angle_2) / 2.f};
```

Comment:
please keep in mind that division is extremely expensive on the GPU - it is even more costly than sin/cos functions, so avoid using divisions at all cost.

for example, always convert `a/2.f` to `a*0.5f` - as the multiplication takes 1 clock cycle to finish and division takes 15 clock cycles or more to complete. 

see 
https://forums.developer.nvidia.com/t/speed-comparison-of-division-compared-to-other-arithmetic-operations-perhaps-something-like-clock-cycles/168371

CUDA even had to use approximated division `__fdividef` to speed up the computation
https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__SINGLE.html#group__CUDA__MATH__INTRINSIC__SINGLE_1gc996beec34f94f6376d0674a6860e107